src/container/srv_container.c

```diff
@@ -45,6 +45,85 @@ static int
 cont_prop_read(struct rdb_tx *tx, struct cont *cont, uint64_t bits,
 	       daos_prop_t **prop_out);
 
+/** Container Property knowledge */
+
+/** Get the redundancy factor from a containers properites. */
+uint32_t
+daos_cont_prop2redunfac(daos_prop_t *props)
+{
+	struct daos_prop_entry *prop =
+		daos_prop_entry_get(props, DAOS_PROP_CO_REDUN_FAC);
+
+	return prop == NULL ? DAOS_PROP_CO_REDUN_RF1 : (uint32_t)prop->dpe_val;
+}
+
+/** Get the redundancy level from a containers properites. */
+uint32_t
+daos_cont_prop2redunlvl(daos_prop_t *props)
+{
+	struct daos_prop_entry *prop =
+		daos_prop_entry_get(props, DAOS_PROP_CO_REDUN_LVL);
+
+	return prop == NULL ? DAOS_PROP_CO_REDUN_NODE : (uint32_t)prop->dpe_val;
+}
+
+/**
+ * This function verifies that the container meets it's redundancy requirements
+ * based on the current pool map. The redundancy requirement is measured
+ * checking a domain level and counting how many downstream failures it has.
+ * if there are more failures for that domain type then allowed restrict
+ * container opening.
+ *
+ * \param pmap  [in]    The pool map referenced by the container.
+ * \param props [in]    The container properties, used to get redundancy factor
+ *                      and level.
+ *
+ * \return	0 if the container meets the requirements, negative error code
+ *		if it does not.
+ */
+static int
+con_verify_redun_req(struct pool_map *pmap, daos_prop_t *props)
+{
+	int num_failed;
+	int rc = 0;
+	int redun_fac = daos_cont_prop2redunfac(props);
+	int redun_lvl = daos_cont_prop2redunlvl(props);
+
+	switch (redun_lvl) {
+	case DAOS_PROP_CO_REDUN_RACK:
+		rc = pool_map_get_failed_cnt(pmap,
+					     PO_COMP_TP_RACK);
+		break;
+	case DAOS_PROP_CO_REDUN_NODE:
+		rc = pool_map_get_failed_cnt(pmap,
+					     PO_COMP_TP_NODE);
+		break;
+	default:
+		return -DER_INVAL;
+	}
+
+	if (rc < 0)
+		return rc;
+
+	num_failed = rc;
+	switch (redun_fac) {
+	case DAOS_PROP_CO_REDUN_RF0:
+		if (num_failed > 0)
+			rc = DER_INVAL;
+		break;
+	case DAOS_PROP_CO_REDUN_RF1:
+		rc = 0;
+		break;
+	case DAOS_PROP_CO_REDUN_RF3:
```

Comment:
shouldn't we check num_failed for these two cases?

Reply:
In **src/include/daos_prop.h** it states,

> RF1: no data protection. scratched data.

I took that to mean there are no check performed for RF1,, is that correct?

Reply:
For the RF_N_ does it follow that they can lose _N_ servers before an error is reported and the container open fails?